src/libraries/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEngine.Unix.cs

```diff
@@ -333,38 +335,32 @@ private void EventLoop()
-                    for (int i = 0; i < numEvents; i++)
+                    foreach (var socketEvent in new ReadOnlySpan<Interop.Sys.SocketEvent>(buffer, numEvents))
                     {
-                        IntPtr handle = buffer[i].Data;
-                        if (handle == ShutdownHandle)
-                        {
-                            shutdown = true;
-                        }
-                        else
+                        IntPtr handle = socketEvent.Data;
+
+                        if (handleToContextMap.TryGetValue(handle, out SocketAsyncContextWrapper contextWrapper) && (context = contextWrapper.Context) != null)
                         {
                             Debug.Assert(handle.ToInt64() < MaxHandles.ToInt64(), $"Unexpected values: handle={handle}, MaxHandles={MaxHandles}");
-                            handleToContextMap.TryGetValue(handle, out SocketAsyncContext? context);
-                            if (context != null)
+
+                            Interop.Sys.SocketEvents events = context.HandleSyncEventsSpeculatively(socketEvent.Events);
+                            if (events != Interop.Sys.SocketEvents.None)
                             {
-                                Interop.Sys.SocketEvents events = buffer[i].Events;
-                                events = context.HandleSyncEventsSpeculatively(events);
-                                if (events != Interop.Sys.SocketEvents.None)
-                                {
-                                    var ev = new SocketIOEvent(context, events);
-                                    eventQueue.Enqueue(ev);
-                                    enqueuedEvent = true;
-
-                                    // This is necessary when the JIT generates unoptimized code (debug builds, live debugging,
-                                    // quick JIT, etc.) to ensure that the context does not remain referenced by this method, as
-                                    // such code may keep the stack location live for longer than necessary
-                                    ev = default;
-                                }
+                                var ev = new SocketIOEvent(context, events);
+                                eventQueue.Enqueue(ev);
+                                enqueuedEvent = true;
 
                                 // This is necessary when the JIT generates unoptimized code (debug builds, live debugging,
                                 // quick JIT, etc.) to ensure that the context does not remain referenced by this method, as
                                 // such code may keep the stack location live for longer than necessary
-                                context = null;
+                                ev = default;
                             }
+
+                            context = null;
+                        }
+                        else if (handle == shutdownHandle)
```

Comment:
Do we have anything guaranteeing that shutdownHandle doesn't end up in the dictionary?  Previously that would have been caught as this check was first, but now it won't.  At a minimum we should have an assert earlier.

Reply:
> Do we ever update the value for an existing key in the dictionary

The dictionary is updated only when a new socket is accepted. So I would say that reads are much more common than writes

Reply:
> Is this nulling out really necessary? We're generally not concerned with nulling out values like this because of debug builds and the like

I wanted to make sure I keep the behaviour the same as in most recent @kouvel PR. I must agree, that I would prefer not to do that and in perfect scenario just avoid all null checks - if `TryGetValue` returns `true` then it's always not null (we control the input values and can ensure that).

Reply:
> Previously that would have been caught as this check was first, but now it won't. At a minimum we should have an assert earlier.

Good point. I am going to check this and add an assert or revert this particular change if we can't ensure it.